CMakeLists.txt

```diff
@@ -303,18 +303,24 @@ if (MSVC)
   else()
     set(EXTRA_FLAGS "")
   endif()
-  if (BUILD_SHARED_LIBS OR ONNX_BUILD_MAIN_LIB)
-    set(ONNX_API_DEFINE "-DONNX_API=__declspec(dllexport)")
+  if (BUILD_SHARED_LIBS)
+    target_compile_definitions(onnx_proto INTERFACE "-DONNX_IMPORT_SYMBOLS")
+    # ONNX_API is set here for compiling onnx_proto itself
+    target_compile_definitions(onnx_proto PRIVATE "-DONNX_API=__declspec(dllexport)")
   else()
-    set(ONNX_API_DEFINE "-DONNX_API=")
+    if (ONNX_BUILD_MAIN_LIB)
+      target_compile_definitions(onnx_proto PUBLIC "-DONNX_EXPORT_SYMBOLS")
```

Comment:
There are two ONNX_BUILD_MAIN_LIB in the original code, a cmake variable (in here) and a compile definition (in onnx_pb.h). 
PUBLIC here is because in the original code ONNX_BUILD_MAIN_LIB cmake variable set the PRIVATE part and the ONNX_BUILD_MAIN_LIB definition in onnx_pb.h set the INTERFACE part. However, I haven't met a use case of both the cmake variable and compile definition.